### DISCUSSION TITLE
Ingress Gateway with Load Balancer support in a mixed environment

### DISCUSSION THREAD

**Discussion by michael-burakov** [istio/istio#52547] [General]
Hi,

My goal is to be able to deploy Kubernetes cluster in any environment, which is usually I don't have control of. So it could be a cloud environment such as AWS, Azure, GCP or an on-premises installation.

So far I had just node ports exposed for services I need to access externally. However, I'm now working to implement an Ingress. As I understand Gateway API is the most flexible alternative.

So my aim for now is to make this work in my test environment, Google Compute Engine VMs where Kubernetes are installed manually (i.e. unmanged, non-GKE cluster).

So I've done some 'getting started' and 'httpbin' tutorials and got it working not without problems. My main problem is, because k8s cluster is unmanaged, 'LoadBalancer' service which Istio Gateway resource is automatically starting, not able to provision an external Load Balancer. Therefore, not "Programmed" and doesn't get IP address assigned.

```
Message:               Assigned to service(s) gateway-istio.istio-ingress.svc.cluster.local:80, but failed to assign to all requested addresses: address pending for hostname "gateway-istio.istio-ingress.svc.cluster.local"
```
Despite the error, routing of HTTP requests from manually provisioned Network Load Balancer  seem to work in 'httpbin' example. However, I'm not so sure if gateway proxy will work correctly without proper IP address configured.

Additionally, it seems not possible to configure predefined ports for auto-generated NodePort of LoadBalancer Service so that external Load Balancer doesn't need to be reconfigured in case Gateway needs to be reporovisioned for any reason.

One solution would be to use MetalLB, as stated in the "Getting Started" guide. However, on the MetalLB documentation page, it is stated that it doesn't support cloud network environments of GCP, AWS and such.

Any suggestions of help is highly appreciated.

**Comment by ilrudie** [comment]
NodePort services for the ingress gateway work perfectly fine but there is just more fiddling to get them setup. It sounds like you're already familiar with that fiddling and would be able to handle it though. I've managed Istio on many clusters that don't support LoadBalancer type services without issue. There is a bit of documentation on this on [istio.io](https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-control/#using-node-ports-of-the-ingress-gateway-service)

**Reply by michael-burakov** [reply]
So, if I understood you correctly, I need to tell Istio Gateway too create a service of type Node Port instead of LoadBalancer?

It looks like I have two options:

* Add annotation `networking.istio.io/service-type` to the gateway yaml when creating it so that service of type NodePort is created.
* Make a manual installation of Ingress Gateway.

First alternative creates Kubernetes Gateway API ingress gateway, as I understand it. It let me specify service type created with an annotation. However, it doesn't look like I can predefine externally exposed port.

The second alternative creates Istio Ingress Gateway (Istio API), right? This option is more flexible as I can choose to specify what and how service is created, but more complex as I need to create the deployment manually as well.

Is it any difference between these options (Kubernetes Gateway API vs Istio API) in terms of internal functionality, or is it only matter of what API is used?

Also a question about the proxy service. Is it nat-aware, i.e. will it modify source IP of outgoing packets. If yes, how does it know what IP does load balancer has when we don't specify it in gateway configuration?

**Reply by ilrudie** [reply]
IMO if you're using nodeport services you are probably stuck with mostly manual gateway creation in general since you're going to be manually configuring a load balancer for it. Just installing an Istio ingress gateway with the helm instructions probably makes a lot of sense for this specific use case although to be fair manual deployment does not hard equal Istio APIs and you can follow these [instructions](https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#manual-deployment) if you want to use gtwapi.

The difference is really mostly preference at this stage although the future is likely gtwapi and users in general may be more familiar/comfortable with gtwapi. To be clear, the future is likely gtwapi is mostly my guess and should not be read as Istio API support is going away. There are no plans to eliminate support for the classic Istio API.

I'm not sure what you mean by "the proxy service" in this case. In my use of Istio for ingress we matched traffic based on host/path and did not worry about source/destination IP.

**Reply by mehyedes** [reply]
Setting the node port should be possible using a configmap referenced by the gateway, as documented in [automated-deployment](https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#automated-deployment).

Check https://github.com/istio/istio/issues/45113#issuecomment-4270548021 for more details.
